src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1006,251 +1083,422 @@ RouterSelectTask(Query *originalQuery, Query *query)
+	ShardInterval *shardInterval = NULL;
+	bool replaceRteWithNullValues = false;
+
+	if (node == NULL)
 	{
 		return false;
 	}
 
-	quals = joinTree->quals;
-	if (quals == NULL)
+	/* want to look at all RTEs, even in subqueries, CTEs and such */
+	if (IsA(node, Query))
 	{
-		return false;
+		return query_tree_walker((Query *) node, UpdateRelationNames, restrictionContext,
+								 QTW_EXAMINE_RTES);
 	}
 
-	/* convert list of expressions into expression tree */
-	if (quals != NULL && IsA(quals, List))
+	if (!IsA(node, RangeTblEntry))
 	{
-		quals = (Node *) make_ands_explicit((List *) quals);
+		return expression_tree_walker(node, UpdateRelationNames, restrictionContext);
 	}
 
-	/*
-	 * Partition column must be used in a simple equality match check and it must be
-	 * place at top level conjustion operator.
-	 */
-	partitionColumnMatchExpression =
-		ColumnMatchExpressionAtTopLevelConjunction(quals, partitionColumn);
 
-	if (!partitionColumnMatchExpression)
+	newRte = (RangeTblEntry *) node;
+
+	if (newRte->rtekind != RTE_RELATION)
 	{
 		return false;
 	}
 
-	/* make sure partition column is used only once in the query */
-	varClauseList = pull_var_clause_default(quals);
-	foreach(varClauseCell, varClauseList)
+	/*
+	 * Search for the restrictions associated with the RTE. There better be
+	 * some, otherwise this query wouldn't be elegible as a router query.
+	 *
+	 * FIXME: We should probably use a hashtable here, to do efficient
+	 * lookup.
+	 */
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
 	{
-		Var *column = (Var *) lfirst(varClauseCell);
-		if (equal(column, partitionColumn))
+		relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+
+		if (GetRTEIdentity(relationRestriction->rte) == GetRTEIdentity(newRte))
 		{
-			partitionColumnReferenceCount++;
+			break;
 		}
+
+		relationRestriction = NULL;
 	}
 
-	if (partitionColumnReferenceCount != 1)
+	replaceRteWithNullValues = (relationRestriction == NULL) ||
+							   relationRestriction->prunedShardIntervalList == NIL;
+
+	if (replaceRteWithNullValues)
 	{
+		ConvertRteToSubqueryWithEmptyResult(newRte);
 		return false;
 	}
 
-	/*
-	 * We need to make sure there is at least one shard for this hash partitioned
-	 * query to be router plannable. We can not prepare a router plan if there
-	 * are no shards.
-	 */
-	shardCount = ShardIntervalCount(distributedTableId);
-	if (shardCount == 0)
+	Assert(relationRestriction != NULL);
+
+	shardIntervalList = relationRestriction->prunedShardIntervalList;
+
+	Assert(list_length(shardIntervalList) == 1);
+	shardInterval = (ShardInterval *) linitial(shardIntervalList);
+
+	shardId = shardInterval->shardId;
+	relationId = shardInterval->relationId;
+	relationName = get_rel_name(relationId);
+	AppendShardIdToName(&relationName, shardId);
+
+	schemaId = get_rel_namespace(relationId);
+	schemaName = get_namespace_name(schemaId);
+
+	/* XXX: why is that a good idea? */
```

Comment:
Verified with @byucesoy that workers expect relation names with schema info now. Removed the removal of "public".